include/drivers/reset.h

```diff
@@ -0,0 +1,192 @@
+ * API template to get the reset status of the device.
+ *
+ * @see reset_status
+ */
+typedef int (*reset_api_status)(const struct device *dev, uint32_t id, uint8_t *status);
+
+/**
+ * API template to put the device in reset state.
+ *
+ * @see reset_assert
+ */
+typedef int (*reset_api_assert)(const struct device *dev, uint32_t id);
+
+/**
+ * API template to take out the device from reset state.
+ *
+ * @see reset_deassert
+ */
+typedef int (*reset_api_deassert)(const struct device *dev, uint32_t id);
+
+/**
+ * API template to reset the device.
+ *
+ * @see reset_toggle
+ */
+typedef int (*reset_api_toggle)(const struct device *dev, uint32_t id);
+
+/**
+ * @brief Reset Controller driver API
+ */
+__subsystem struct reset_driver_api {
+	reset_api_status status;
+	reset_api_assert assert;
+	reset_api_deassert deassert;
+	reset_api_toggle toggle;
+};
+
+/** @endcond */
+
+/**
+ * @brief Get the reset status
+ *
+ * This function returns the reset status of the device.
+ *
+ * @param reset Reset controller device configuration.
+ * @param status Where to write the reset status.
+ *
+ * @return 0 on success, negative code on error.
+ */
+__syscall int reset_status(const struct reset_dev_config *reset, uint8_t *status);
+
+static inline int z_impl_reset_status(const struct reset_dev_config *reset, uint8_t *status)
+{
+	const struct reset_driver_api *api = (const struct reset_driver_api *)reset->dev->api;
+
+	if (!api) {
+		return -ENOTSUP;
+	}
+
+	return api->status(reset->dev, reset->id, status);
+}
+
+/**
+ * @brief Put the device in reset state
+ *
+ * This function sets/cleares the reset bits of the device,
+ * depending on the logic level (active-high/active-low).
+ *
+ * @param reset Reset controller device configuration.
+ *
+ * @return 0 on success, negative code on error.
+ */
+__syscall int reset_assert(const struct reset_dev_config *reset);
+
+static inline int z_impl_reset_assert(const struct reset_dev_config *reset)
+{
+	const struct reset_driver_api *api = (const struct reset_driver_api *)reset->dev->api;
+
+	if (!api) {
+		return -ENOTSUP;
+	}
+
+	return api->assert(reset->dev, reset->id);
+}
+
+/**
+ * @brief Take out the device from reset state.
+ *
+ * This function sets/cleares the reset bits of the device,
+ * depending on the logic level (active-low/active-high).
+ *
+ * @param reset Reset controller device configuration.
+ *
+ * @return 0 on success, negative code on error.
+ */
+__syscall int reset_deassert(const struct reset_dev_config *reset);
+
+static inline int z_impl_reset_deassert(const struct reset_dev_config *reset)
+{
+	const struct reset_driver_api *api = (const struct reset_driver_api *)reset->dev->api;
+
+	if (!api) {
+		return -ENOTSUP;
+	}
+
+	return api->deassert(reset->dev, reset->id);
+}
+
+/**
+ * @brief Reset the device.
+ *
+ * This function performs reset for a device (assert + deassert).
+ *
+ * @param reset Reset controller device configuration.
+ *
+ * @return 0 on success, negative code on error.
```

Comment:
I'd suggest to use `@retval`, since some error codes like `-ENOSYS` can be explained better:

```
@retval 0 On success
@retval -ENOSYS If the functionality is not implemented by the device
@retval -errno Other negative errno in case of failure
```